src/Symfony/Component/Security/Core/Encoder/ChainPasswordEncoder.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Encoder;
+
+/**
+ * Hashes passwords using the best available encoder.
+ * Validates them using a chain of encoders.
+ *
+ * /!\ Don't put a PlaintextPasswordEncoder in the list as that'd mean a leaked hash
+ * could be used to authenticate successfully without knowing the cleartext password.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class ChainPasswordEncoder extends BasePasswordEncoder implements SelfSaltingEncoderInterface
```

Comment:
This name is bit misleading to me. It sounds like multiple encoders are asked to encode the password, but there is actually just one that will encode the password, but multiple (the chain) that will state if the password is valid.

Reply:
Chaining logic belongs to each domain. Here, the practice of using several encoders is not documented in the literature: it would give a false sense of security when a single good algo is the correct bullet.
I'm fine with the current name. Do we need another name? If yes which one?

Reply:
I think we're running into this problem because ideally the encoder and the validator would be separate interfaces (Interface Seggregation principle, one for encoding, another for validation), but pragmatically it's nicer to couple them into one for symmetry.

Anyway, maybe something like `BackwardCompatiblePasswordEncoder`? Or `OldPasswordSchemeAwarePasswordEncoder`?

Reply:
That would suggest that out of all the `PasswordEncoder`s, this one is the best, which is not what we're trying to say :wink:

Reply:
Yes it is :)
It's the best because it enables migrations, and these are the best :)

Reply:
But what if we come up with another idea, and that is actually better, what will we call that? Can't call it the best, that's already taken, but what is better than the best? See where I'm going with this? :wink:

Reply:
MigratingPasswordEncoder

Reply:
I like that :+1: